WordPress/Classes/Login/LoginWithUrlView.swift

```diff
@@ -0,0 +1,158 @@
+            }
+
+            isLoading = false
+        }
+    }
+}
+
+private extension LoginClient.LoginClientError {
+
+    var errorMessage: String? {
+        switch self {
+        case let .invalidSiteAddress(error):
+            return error.errorMessage
+        case .missingLoginUrl:
+            return NSLocalizedString("addSite.selfHosted.noLoginUrlFound", value: "Application Password authentication needs to be enabled on the WordPress site.", comment: "Error message shown when application-password authentication is disabled on a self-hosted WordPress site")
+        case let .authenticationFailure(error):
+            if error.code == .canceledLogin {
+                return nil
+            }
+            fallthrough
+        case .invalidApplicationPasswordCallback:
+            return NSLocalizedString("addSite.selfHosted.authenticationFailed", value: "Cannot login using Application Password authentication.", comment: "Error message shown when an receiving an invalid application-password authentication result from a self-hosted WordPress site")
+        case .loadingSiteInfoFailure:
+            return NSLocalizedString("addSite.selfHosted.loadingSiteInfoFailure", value: "Cannot load the WordPress site details", comment: "Error message shown when failing to load details from a self-hosted WordPress site")
+        case .savingSiteFailure:
+            return NSLocalizedString("addSite.selfHosted.savingSiteFailure", value: "Cannot save the WordPress site, please try again later.", comment: "Error message shown when failing to save a self-hosted site to user's device")
+        case .unknown:
+            return NSLocalizedString("addSite.selfHosted.unknownError", value: "Something went wrong. Please try again.", comment: "Error message when an unknown error occurred when adding a self-hosted site")
+        }
+    }
+
+}
+
+// MARK: - WordPressAPI helpers
+
+private extension UrlDiscoveryError {
+    var errorMessage: String? {
+        let errors: [UrlDiscoveryAttemptError]
+
+        switch self {
+        case let .UrlDiscoveryFailed(attempts):
+            errors = attempts
+                .compactMap {
+                    switch $0 {
+                    case let .failure(failure):
+                        return failure
+                    case .success:
+                        return nil
+                    }
+                }
+        }
+
+        let notWordPressSite = errors.contains {
+            switch $0 {
+            case .fetchApiRootUrlFailed, .fetchApiDetailsFailed:
+                return true
+            case .failedToParseSiteUrl:
+                return false
+            }
+        }
+
+        if notWordPressSite {
+            return NSLocalizedString("addSite.restApiNotAvailable", value: "The site at this address is not a WordPress site. For us to connect to it, the site must use WordPress.", comment: "Error message shown a URL does not point to an existing site.")
```

Comment:
(As discussed on Slack) this should be pushed down to the shared layer, if possible

Reply:
I didn't do it in this PR because I feel like that'll be a whole little project on its own.

We'll need to set up translation pipeline for the library, including Swift and Kotlin code to use a shared localized string keys, tooling for extracting those localized string keys out for translation, importing translation results back to Swift and Kotlin libraries (which may affects how we release the library).

What do you think leaving these code here for now, and updating them after we have a working localization supports in wordpress-rs?

Reply:
> What do you think leaving these code here for now, and updating them after we have a working localization supports in wordpress-rs?

Sure – I can live with that